test/server/server_test.cc

```diff
@@ -134,6 +134,93 @@ TEST_F(RunHelperTest, ShutdownBeforeInitManagerInit) {
+                     file_system, std::move(process_context)) {}
+
+  Init::Manager& initManager() override { return init_manager_; }
+};
+
+// Class creates minimally viable server instance for testing and the initManager managed by it is
+// stuck in Initializing.
+class InitializingServerInstanceImplTest
+    : public testing::TestWithParam<Network::Address::IpVersion> {
+protected:
+  InitializingServerInstanceImplTest() : version_(GetParam()) {}
+
+  void initialize(const std::string& bootstrap_path) {
+    if (bootstrap_path.empty()) {
+      options_.config_path_ = TestEnvironment::temporaryFileSubstitute(
+          "test/config/integration/server.json", {{"upstream_0", 0}, {"upstream_1", 0}}, version_);
+    } else {
+      options_.config_path_ = TestEnvironment::temporaryFileSubstitute(
+          bootstrap_path, {{"upstream_0", 0}, {"upstream_1", 0}}, version_);
+    }
+    thread_local_ = std::make_unique<ThreadLocal::InstanceImpl>();
+    if (process_object_ != nullptr) {
+      process_context_ = std::make_unique<ProcessContextImpl>(*process_object_);
+    }
+    server_ = std::make_unique<InitializingInstanceImpl>(
+        options_, test_time_.timeSystem(),
+        Network::Address::InstanceConstSharedPtr(new Network::Address::Ipv4Instance("127.0.0.1")),
+        hooks_, restart_, stats_store_, fakelock_, component_factory_,
+        std::make_unique<NiceMock<Runtime::MockRandomGenerator>>(), *thread_local_,
+        Thread::threadFactoryForTest(), Filesystem::fileSystemForTest(),
+        std::move(process_context_));
+
+    EXPECT_TRUE(server_->api().fileSystem().fileExists("/dev/null"));
+  }
+  Network::Address::IpVersion version_;
+  testing::NiceMock<MockOptions> options_;
+  DefaultListenerHooks hooks_;
+  testing::NiceMock<MockHotRestart> restart_;
+  std::unique_ptr<ThreadLocal::InstanceImpl> thread_local_;
+  Stats::TestIsolatedStoreImpl stats_store_;
+  Thread::MutexBasicLockable fakelock_;
+  TestComponentFactory component_factory_;
+  DangerousDeprecatedTestTime test_time_;
+  ProcessObject* process_object_ = nullptr;
+  std::unique_ptr<ProcessContextImpl> process_context_;
+  std::unique_ptr<InstanceImpl> server_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, InitializingServerInstanceImplTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// Validates that server stats are created even if the server is stuck with initialization.
+TEST_P(InitializingServerInstanceImplTest, StatsFlushWhenServerIsStillInitializing) {
+  options_.hot_restart_epoch_ = 3;
+
+  EXPECT_NO_THROW(initialize("test/server/empty_bootstrap.yaml"));
+
+  // Validate that server is still in initializing state but stats are created and flushed.
+  EXPECT_EQ(Init::Manager::State::Initializing, server_->initManager().state());
+  EXPECT_EQ(1L, TestUtility::findGauge(stats_store_, "server.live")->value());
```

Comment:
I'm not sure that this actually tests the change? Wouldn't these stats exist in the store even modulo the change? It seems like maybe you need a custom sink to make sure you get the stats?

Reply:
Yeah. I actually meant to test `flushStats` directly without writing a custom sink. I have corrected the test now. PTAL.